# Read mapping and SNP calling pipeline
The file 'run.fastp.sh' contains the script used to run fastp and assess the quality of each sample. Additionally, the file 'run.mapping.sh' contains the script used to map cleaned reads to the reference genome and call SNPs using Stacks. The file 'pheucticus.popmap.no.oregon.txt' is the only input needed for these scripts other than the raw fastq file for each sample, which can be downloaded from NCBI. With those inputs you could recreate from scratch our unfiltered SNP dataset, which is input for the next step in the protocol, [SNP filtering](https://devonderaad.github.io/grosbeak.rad.hybridtransect/analyses/grosbeak.filtering.html).
